### PR TITLE
Support overloading various operations

### PR DESCRIPTION
**Pull request by esohns**
- support overloading ACE_Module
- support overloading ACE_Message_Queue_Iterator(s) (this allows adding support for grabbing the lock just once per complete iteration)
- support overloading SSL_SOCK_Connector
- support ::CancelIoEx on Windows >= Vista

### CODE REVIEW

File: ACE/ace/WIN32_Asynch_IO.cpp [hunk 1]
```diff
@@ -168,25 +168,25 @@ ACE_WIN32_Asynch_Operation::open (const ACE_Handler::Proxy_Ptr &handler_proxy,
...
 {
-#if defined (ACE_HAS_WIN32_OVERLAPPED_IO)
+#if defined (ACE_HAS_CANCEL_IO)
   // All I/O operations that are canceled will complete with the error
   // ERROR_OPERATION_ABORTED. All completion notifications for the I/O
   // operations will occur normally.
 
-  // @@ This API returns 0 on failure. So, I am returning -1 in that
-  //    case. Is that right? (Alex).
+#if (_WIN32_WINNT < _WIN32_WINNT_VISTA)
```
Comment (jwillemsen) [inline ACE/ace/WIN32_Asynch_IO.cpp:176]: Recommend to use 0x0600, older SDKs probably don't have the _WIN32_WINNT_VISTA define

``` suggestion
#if (_WIN32_WINNT < 0x0600)
```

**Comment by jwillemsen** [comment]
Looks there is a fuzz error
```
Error: ./ACE/ace/WIN32_Asynch_IO.cpp:179: NULL found
```

**Comment by jwillemsen** [comment]
Is from http://bugzilla.dre.vanderbilt.edu/show_bug.cgi?id=4220

File: ACE/ace/Module.h [hunk 2]
```diff
@@ -59,6 +59,8 @@ class ACE_Export ACE_Module_Base
     /// Indicates that close() should not delete any Tasks.
     M_DELETE_NONE = 4
   };
+
+  inline virtual ~ACE_Module_Base (void) {}
```
Comment (mitza-oci) [inline ACE/ace/Module.h:63]: ACE-guidelines.html says this shouldn't be inline ("If a class has any virtual functions..."). Not sure it's still a concern, but it does make sense that for efficient code generation / linking there should be a single object file for the vtable/typeinfo.

File: ACE/ace/WIN32_Asynch_IO.cpp [hunk 3]
```diff
@@ -168,25 +168,25 @@ ACE_WIN32_Asynch_Operation::open (const ACE_Handler::Proxy_Ptr &handler_proxy,
 int
 ACE_WIN32_Asynch_Operation::cancel (void)
 {
-#if defined (ACE_HAS_WIN32_OVERLAPPED_IO)
+#if defined (ACE_HAS_CANCEL_IO)
```
Comment (mitza-oci) [inline ACE/ace/WIN32_Asynch_IO.cpp:171]: split to a separate PR?

File: ACE/ace/Message_Queue_T.h [hunk 4]
```diff
@@ -669,12 +669,13 @@ class ACE_Message_Queue_Iterator
   /// Declare the dynamic allocation hooks.
   ACE_ALLOC_HOOK_DECLARE;
 
+protected:
+  /// Keeps track of how far we've advanced...
+  ACE_Message_Block *curr_;
```
Comment (mitza-oci) [inline ACE/ace/Message_Queue_T.h:674]: Since this changes the declaration order of the data members, the order that those members appear in the constructor's member initializer list needs to change as well.

File: ACE/ace/Module.h (hunk 2 above)
Comment (jwillemsen) [reply ACE/ace/Module.h:63]: Yes, you are right

Comment (esohns) [reply ACE/ace/Module.h]: that didn't work (on Win32) for some reason, removed this change for now

File: ACE/ace/Message_Queue_T.h (hunk 4 above)
Comment (esohns) [reply ACE/ace/Message_Queue_T.h:674]: done

**Review by jwillemsen — APPROVED** [review]

**Comment by jwillemsen** [comment]
Thanks!

**Comment by jwillemsen** [comment]
Looks this change still triggers some new warnings

```
/home/build/jenkins/workspace/sles150_ace_i686_debug/BUILD/DOC_ROOT/ACE/ace/Message_Queue_T.cpp: In function ‘int iterator_test()’:
/home/build/jenkins/workspace/sles150_ace_i686_debug/BUILD/DOC_ROOT/ACE/ace/Message_Queue_T.cpp:746:19: warning: ‘iterator.ACE_Message_Queue_Reverse_Iterator<ACE_NULL_SYNCH>::queue_’ may be used uninitialized in this function [-Wmaybe-uninitialized]
: curr_ (queue_.tail_)
~~~~~~~^~~~~
/home/build/jenkins/workspace/sles150_ace_i686_debug/BUILD/DOC_ROOT/ACE/tests/Message_Queue_Test.cpp:368:22: note: ‘iterator.ACE_Message_Queue_Reverse_Iterator<ACE_NULL_SYNCH>::queue_’ was declared here
REVERSE_ITERATOR iterator (queue);
^~~~~~~~
```
